### PR TITLE
Dev suite code for QL

### PR DESCRIPTION
**Pull request by profxj**
This pairs with https://github.com/pypeit/PypeIt/pull/1511 

The main change is a new and fixed data model
for specifying tests in test_setups.py.

That and a series of new QL tests.

Also, we are now going to have one DEIMOS reduce test
be run on *all* the detectors.  Yup, a bit slow..

### CODE REVIEW

File: test_scripts/pypeit_tests.py [hunk 1]
```diff
@@ -465,51 +465,62 @@ class PypeItQuickLookTest(PypeItTest):
...
         for option in self.options:
             if self.options[option] is None:
                 command_line += [option]
+            elif self.options[option] in ['USE_MASTERS_DIR', 
+                                          'USE_CALIB_DIR']:
+                idir = os.path.join(self.redux_dir,
+                    self.setup.instr, self.setup.name)
+                # Masters?
+                if self.options[option] == 'USE_MASTERS_DIR':
+                    # Crazy hack for pypeit_setup
```
Comment (badpandabear) [inline test_scripts/pypeit_tests.py:516]: I think you can avoid this hack by using self.setup.rdxdir, which is updated in PypeItSetupTest to include the "instrument"_A directory.  You might have to run os.path.abspath on it though.

File: test_scripts/test_setups.py [hunk 2]
```diff
@@ -100,140 +99,237 @@ class TestPhase(Enum):
...
+}
+
+# Tests for full reductions
+_reduce_setups = {}
+for instr in all_setups:
+    _reduce_setups[instr] = {}
+    for setup in all_setups[instr]:
+        _reduce_setups[instr][setup] = [{}]
+
+#all_setups = copy.deepcopy(reduce_setups)
```
Comment (badpandabear) [inline test_scripts/test_setups.py:159]: Remove commented out code

File: test_scripts/test_setups.py [hunk 3]
```diff
@@ -100,140 +99,237 @@ class TestPhase(Enum):
...
+#all_setups['keck_deimos'].append('QL')
+
+# If you add a test here and it is in QL
+#   you will need may modify the QuickLookTest class
```
Comment (badpandabear) [inline test_scripts/test_setups.py:163]: I think you can avoid having to modify the QuickLookTest class, see my note in pypeit_tests.py

File: test_scripts/test_main.py [hunk 4]
```diff
@@ -997,61 +997,70 @@ def build_test_setup(pargs, instr, setup_name, flg_reduce, flg_after, flg_ql):
...
-                key = setup.instr
 
-        elif setup.key in test_descr['setups']:
-            # Either a dict or list of setups
-            key = setup.key
-        else:
+        # Check instruments
+        if setup.instr not in test_descr['setups'].keys():
+            continue
+#            if isinstance(test_descr['setups'][setup.instr], list):
```
Comment (badpandabear) [inline test_scripts/test_main.py:1004]: Remove commented out code

**Review by badpandabear — APPROVED** [review]
I really like the changes you made to test_setups.py and build_test_setup!

File: test_scripts/test_main.py [hunk 5]
```diff
@@ -997,61 +997,70 @@ def build_test_setup(pargs, instr, setup_name, flg_reduce, flg_after, flg_ql):
...
-            else:
-                # A dict mapping to arguments that apply to all setups for this instrument
-                key = setup.instr
 
-        elif setup.key in test_descr['setups']:
-            # Either a dict or list of setups
-            key = setup.key
-        else:
+        # Check instruments
+        if setup.instr not in test_descr['setups'].keys():
```
[the remark this answers is not in the document]
Comment (profxj) [reply test_scripts/test_main.py:1002]: done

File: test_scripts/test_main.py [hunk 6]
```diff
@@ -997,61 +997,70 @@ def build_test_setup(pargs, instr, setup_name, flg_reduce, flg_after, flg_ql):
...
+#                key = setup.instr
+#
+#        elif setup.key in test_descr['setups']:
+#            # Either a dict or list of setups
+#            key = setup.key
+#        else:
+#            continue
+
+        # Check setup
+        if setup_name not in test_descr['setups'][setup.instr].keys():
```
[the remark this answers is not in the document]
Comment (profxj) [reply test_scripts/test_main.py:1021]: done

File: test_scripts/test_main.py [hunk 7]
```diff
@@ -997,61 +997,70 @@ def build_test_setup(pargs, instr, setup_name, flg_reduce, flg_after, flg_ql):
...
+#        if isinstance(test_descr['setups'], dict) and key is not None:
+#            kwargs = test_descr['setups'][key]
+#        else:
+#            kwargs = dict()
+
+        for kwargs in test_descr['setups'][setup.instr][setup_name]:
+            # Create the test, this will also run any prep_only steps in the
+            # __init__ method
+            try:
+                #test = test_descr['factory'](setup, pargs, **kwargs)
```
[the remark this answers is not in the document]
Comment (profxj) [reply test_scripts/test_main.py:1035]: gone, along with a number of other old code and extraneous comments

File: test_scripts/pypeit_tests.py (hunk 1 above)
Comment (profxj) [reply test_scripts/pypeit_tests.py:516]: nice! done

File: test_scripts/test_setups.py (hunk 2 above)
Comment (profxj) [reply test_scripts/test_setups.py:159]: removed

File: test_scripts/test_setups.py (hunk 3 above)
Comment (profxj) [reply test_scripts/test_setups.py:163]: removed

File: test_scripts/test_main.py (hunk 4 above)
Comment (profxj) [reply test_scripts/test_main.py:1004]: removed

**Pull request merged**